newm-server/src/main/kotlin/io/newm/server/features/song/repo/SongRepositoryImpl.kt

```diff
@@ -382,30 +388,62 @@ internal class SongRepositoryImpl(
+        val usdAdaExchangeRate = cardanoRepository.queryAdaUSDPrice().toBigInteger()
+
+        val sendTokenFee = (numberOfCollaborators * minUtxo)
+        val mintCostLovelace = mintCostBase + sendTokenFee
+
+        return calculateMintPaymentResponse(
+            dspPriceUsd,
+            usdAdaExchangeRate,
+            mintCostLovelace,
+            sendTokenFee,
+        ).also {
+            // Save the total cost to distribute and mint to the database
+            val totalCostLovelace = BigDecimal(it.adaPrice!!).movePointRight(6).toLong()
+            update(songId, Song(mintCostLovelace = totalCostLovelace))
+        }
+    }
 
-        // Save the mint cost to the database
-        update(songId, Song(mintCostLovelace = mintCostLovelace))
+    @VisibleForTesting
+    internal fun calculateMintPaymentResponse(
+        dspPriceUsd: Long,
+        usdAdaExchangeRate: BigInteger,
+        mintCostLovelace: Long,
+        sendTokenFee: Long,
+    ): MintPaymentResponse {
+        val changeAmountLovelace = 1000000L // 1 ada
+        val dspPriceLovelace =
+            dspPriceUsd.toBigDecimal().divide(usdAdaExchangeRate.toBigDecimal(), 6, RoundingMode.CEILING)
+                .times(1000000.toBigDecimal()).toBigInteger()
 
         // usdPrice does not include the extra changeAmountLovelace that we request the wallet to provide as it
         // is returned to the user.
-        val usdPrice = (
-            cardanoRepository.queryAdaUSDPrice()
-                .toBigInteger() * mintCostLovelace.toBigInteger()
-            ) / 1000000.toBigInteger()
-
-        return Pair(
-            // we send an extra changeAmountLovelace to ensure we have enough ada to cover a return utxo
-            CborInteger.create(mintCostLovelace + changeAmountLovelace).toCborByteArray().toHexString(),
-            usdPrice
+        val usdPrice =
+            usdAdaExchangeRate * (mintCostLovelace.toBigInteger() + dspPriceLovelace) / 1000000.toBigInteger()
+
+        val sendTokenFeeUsd = usdAdaExchangeRate * sendTokenFee.toBigInteger() / 1000000.toBigInteger()
+
+        return MintPaymentResponse(
+            cborHex = // we send an extra changeAmountLovelace to ensure we have enough ada to cover a return utxo
+            CborInteger.create(mintCostLovelace + dspPriceLovelace.toLong() + changeAmountLovelace).toCborByteArray()
+                .toHexString(),
+            adaPrice = (mintCostLovelace.toBigInteger() + dspPriceLovelace).toAda().toPlainString(),
+            usdPrice = usdPrice.toAda().toPlainString(),
+            dspPriceAda = dspPriceLovelace.toAda().toPlainString(),
+            dspPriceUsd = dspPriceUsd.toBigInteger().toAda().toPlainString(),
+            sendTokenFeeAda = sendTokenFee.toBigInteger().toAda().toPlainString(),
+            sendTokenFeeUsd = sendTokenFeeUsd.toAda().toPlainString(),
+            usdAdaExchangeRate = usdAdaExchangeRate.toAda().toPlainString(),
```

Comment:
Nit: Maybe is worth creating a  `toAdaString()` extension that combines `toAda().toPlainString()`